### PR TITLE
Deprecate mutable data type

### PR DESCRIPTION
**Pull request by madmongo1**
Refactor dynamic buffers to ensure conformance with DynamicBuffers_v1 interface

### CODE REVIEW

File: include/boost/beast/core/detail/buffer_utils.hpp [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+auto
+truncate(
+    Buffer &target,
+    std::size_t max_len)
+-> Buffer &
+{
+    target = Buffer(target.data(),
+                    (std::min)(max_len, target.size()));
+    return target;
+}
```
Comment (vinniefalco) [inline include/boost/beast/core/detail/buffer_utils.hpp]: This goes in `boost/beast/core/detail/buffer.hpp` and it should be two inline function overloads, no templates

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 2]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-    end() const;
-
-private:
-    friend class buffers_adaptor;
-
-    mutable_buffers_type(
-            buffers_adaptor const& b)
-        : b_(&b)
+    struct iterator_category
+        : std::bidirectional_iterator_tag
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: why not
```
using iterator_category = std::bidirectional_iterator_tag;
```
?

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 3]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
     }
 }
 
+template<class MutableBufferSequence>
+auto
+buffers_adaptor<MutableBufferSequence>::
+make_subrange(std::size_t pos, std::size_t n)
+-> subrange<true>
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]:
```
template<class MutableBufferSequence>
auto
buffers_adaptor<MutableBufferSequence>::
make_subrange(
    std::size_t pos, std::size_t n) ->
        subrange<true>
```

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 4]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
+        subrange const&) = default;
 #endif
 
-    template<bool isMutable_ = isMutable, class =
-        typename std::enable_if<! isMutable_>::type>
-    readable_bytes(
-        readable_bytes<true> const& other) noexcept
-        : b_(other.b_)
+    // allow conversion from mutable to const
+    template<bool thisIsMutable = isMutable,
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: I usually just throw an underscore in the mix (or remove it), e.g. `isMutable_`

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 5]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
+        typename std::enable_if<!thisIsMutable>::type * = nullptr>
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: I would put `typename` on the previous line, since it adds less than one bit of information. Not a showstopper though.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 6]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-        return b_ == other.b_ && it_ == other.it_;
-    }
-
-    bool
-    operator!=(const_iterator const& other) const
-    {
-        return !(*this == other);
-    }
+    iter_type
+    first() const;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: `get_buffer`, `first`, `last`, `first_offset`, and `last_size` should not be public, since they are not public interfaces.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 7]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
 
 //------------------------------------------------------------------------------
 
 template<class MutableBufferSequence>
-class buffers_adaptor<MutableBufferSequence>::
-mutable_buffers_type
+template<bool isMutable>
+struct buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: Formatting
```
template<bool isMutable>
struct buffers_adaptor<MutableBufferSequence>::
    subrange<isMutable>::
    iterator
```

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 8]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-    buffers_adaptor const* b_ = nullptr;
-
-public:
-    using value_type = net::mutable_buffer;
-    using pointer = value_type const*;
-    using reference = value_type;
+    };
+    using value_type = typename buffers_adaptor<MutableBufferSequence>::
+    template subrange<isMutable>::
+    value_type;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: Continuation lines have to be indented:
```
    using value_type = typename
        buffers_adaptor<MutableBufferSequence>:: template
        subrange<isMutable>::value_type;
```

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 9]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
+    using reference = typename std::add_lvalue_reference<value_type>::type;
+    using pointer = typename std::add_pointer<value_type>::type;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: There's no need for using `add_pointer` or `add_lvalue_reference` why did you do that?

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 10]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
     using difference_type = std::ptrdiff_t;
-    using iterator_category =
-        std::bidirectional_iterator_tag;
 
-    const_iterator() = default;
-    const_iterator(const_iterator const& other) = default;
-    const_iterator& operator=(const_iterator const& other) = default;
+    iterator(
+        typename buffers_adaptor<MutableBufferSequence>::
+        template subrange<isMutable> const *parent,
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: Continuation lines should be indented
```
    iterator(typename
        buffers_adaptor<MutableBufferSequence>::template
            subrange<isMutable> const *parent,
        iter_type it);
```

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 11]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-        return value_type{b.data(),
-            it_ == std::prev(b_->end_) ?
-                b_->out_end_ : b.size()} +
-                    (it_ == b_->out_ ? b_->out_pos_ : 0);
-    }
+    iterator &
+    operator++();
 
-    pointer
-    operator->() const = delete;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: Why did you remove this?

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 12]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
+    std::size_t
+    first_offset() const;
 
-    const_iterator&
-    operator++()
-    {
-        ++it_;
-        return *this;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: Why are the function definitions moved?

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 13]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+iterator::
+operator*() const
+-> value_type
+{
+    return parent_->get_buffer(it_);
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: There's no need for `get_buffer` since it is only called from this place, might as well just put the code here.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 14]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+get_buffer(iter_type it) const
+-> value_type
+{
+    auto result = *it;
+
+    if (it == first_)
+        result += first_offset_;
+
+    if (std::next(it) == last_)
+        detail::truncate(result, last_size_);
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: There's no need to have a separate `truncate` function, since it is only called from here. And even if it wasn't, the operation is so trivial that it is not deserving of being elevated to its own function.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 15]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-        readable_bytes<true> const& other) noexcept
-        : b_(other.b_)
+    // allow conversion from mutable to const
+    template<bool thisIsMutable = isMutable,
+        typename std::enable_if<!thisIsMutable>::type * = nullptr>
+    subrange(subrange<true> const &other)
+        : first_(other.first())
+        , last_(other.last())
+        , first_offset_(other.first_offset())
+        , last_size_(other.last_size())
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: These member functions `first()`, `last()`, `first_offset()`, and `last_size()` seem unnecessary, you could just access the data members. You will have to make the class template a friend first of course.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 16]
```diff
@@ -34,332 +35,162 @@ namespace beast {
...
-    friend class mutable_buffers_type;
 
-    const_iterator(buffers_adaptor const& b,
-            iter_type iter)
-        : it_(iter)
-        , b_(&b)
-    {
-    }
+    buffers_adaptor<MutableBufferSequence>::
+    subrange<isMutable> const *parent_;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: continuation lines should be indented
```
    buffers_adaptor<MutableBufferSequence>::
        subrange<isMutable> const *parent_;
```
I'm not a fan of "parent" but it isn't a showstopper. I would have just used `b_`.

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 17]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+adjust(
+    std::size_t pos,
+    std::size_t n)
+-> void
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: We should never see `-> void`

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 18]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+{
+    if (n == 0)
+        last_ = first_;
+
+    // note: fallthrough
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: cryptic

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 19]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+        if (adjust >= first_->size())
+        {
+            ++first_;
+            first_offset_ = 0;
+            pos -= adjust;
+        }
+        else
+        {
+            first_offset_ = adjust;
+            pos = 0;
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: you could `break` and skip the loop condition...

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 20]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+    // note: fallthrough
+
+    if (first_ == last_)
+    {
+        first_offset_ = 0;
+        last_size_ = 0;
+        return;
+    }
+
+    auto last_element = [this](iter_type iter) {
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: `auto is_last = ...`

File: include/boost/beast/core/impl/buffers_adaptor.hpp [hunk 21]
```diff
@@ -601,6 +437,329 @@ consume(std::size_t n) noexcept
...
+    }
+}
+
+
+template<class MutableBufferSequence>
+template<bool isMutable>
+auto
+buffers_adaptor<MutableBufferSequence>::
+subrange<isMutable>::
+begin() const -> iterator
```
Comment (vinniefalco) [inline include/boost/beast/core/impl/buffers_adaptor.hpp]: I do like return types on their own line, whether auto or not
```
template<class MutableBufferSequence>
template<bool isMutable>
auto
buffers_adaptor<MutableBufferSequence>::
subrange<isMutable>::
begin() const ->
    iterator
```

File: test/beast/core/buffers_adaptor_subrange.cpp [hunk 22]
```diff
@@ -0,0 +1,134 @@
...
+        std::size_t n = (std::numeric_limits<std::size_t>::max)())
+    -> typename buffers_adaptor<MutableBufferSequence>::const_buffers_type
+    {
+        return adaptor.make_subrange(pos, n);
+    }
+};
+
+namespace detail {
+
+class buffers_adaptor_subrange_test
```
Comment (vinniefalco) [inline test/beast/core/buffers_adaptor_subrange.cpp]: This all needs to go in buffers_adaptor.cpp not this new weird file which should not exist

File: test/beast/core/test_buffer.hpp [hunk 23]
```diff
@@ -360,21 +360,21 @@ test_mutable_dynamic_buffer(
 
 /** Test an instance of a dynamic buffer or mutable dynamic buffer.
```
Comment (vinniefalco) [inline test/beast/core/test_buffer.hpp]: There should be no such thing as "mutable dynamic buffer". All of Beast's dynamic buffers should support both `const` and non-`const` `data()`. You might need a separate commit for that.

File: test/beast/core/test_helpers.hpp [hunk 24]
```diff
@@ -0,0 +1,218 @@
...
+        return iterator();
+    }
+private:
+    std::size_t start_;
+    std::size_t limit_;
+    std::size_t interval_;
+};
+
+
+} // test_heplers
```
Comment (vinniefalco) [inline test/beast/core/test_helpers.hpp]: No I've decided I don't want to start adding this to every test, we need to think of a better way.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 11 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: the iterator is now a subtype of subrange. It wasn't that I deleted this specifically, I just didn't write it. What is the purpose of this deleted function?

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 12 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: the enture class has been replaced.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 13 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: This boils down to our fundamentally different views on nested classes. I disagree, but will change it.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 14 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: strongly disagree.
My reasoning is that one logical operation deserves its own function and name as this aids higher level reasoning about intent.
Again, will change.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 16 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: I think `b_` is un-necessarily cryptic - I speak as someone who used to be quite happy being limted to 6 characters for label names in 6809 assembler. Even then we'd try to spll wt mnngfl nms.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 19 above)
Comment (madmongo1) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: This would have no effect on the quality of gcc's code, and would for me make the loop more difficult to reason about as it's another exit point.

https://godbolt.org/z/yNsgjF

nevertheless will add

File: test/beast/core/test_buffer.hpp (hunk 23 above)
Comment (madmongo1) [reply test/beast/core/test_buffer.hpp]: I don't understand what you mean.
The code comment about was not written by me.

File: include/boost/beast/core/impl/buffers_adaptor.hpp (hunk 11 above)
Comment (vinniefalco) [reply include/boost/beast/core/impl/buffers_adaptor.hpp]: to inform the user that the omission is intended (`operator->` is unimplementable)

File: test/beast/core/test_buffer.hpp (hunk 23 above)
Comment (vinniefalco) [reply test/beast/core/test_buffer.hpp]: Yep I know, lets discuss it

File: test/beast/core/buffers_adaptor.cpp [hunk 25]
```diff
@@ -85,6 +113,69 @@ class buffers_adaptor_test : public unit_test::suite
...
+            using maybe_mutable =
+            typename std::conditional<
+                isMutable,
+                buffers_adaptor<std::vector<net::mutable_buffer>>&,
+                buffers_adaptor<std::vector<net::mutable_buffer>> const&>::type;
+
+            auto sub = buffers_adaptor_test_hook::make_subrange(static_cast<maybe_mutable>(adapter));
+            BEAST_EXPECTS(typeid(typename decltype(sub)::value_type) == typeid(value_type), func);
+
+            auto as_string = [](decltype(sub) const &sr) {
```
Comment (vinniefalco) [inline test/beast/core/buffers_adaptor.cpp]: What's wrong with calling `beast::buffers_to_string`? It is even more efficient than this function (since it reserves the space first)
https://www.boost.org/doc/libs/develop/libs/beast/doc/html/beast/ref/boost__beast__buffers_to_string.html

File: test/beast/core/buffers_adaptor.cpp [hunk 26]
```diff
@@ -85,6 +113,69 @@ class buffers_adaptor_test : public unit_test::suite
...
+                std::string result;
+                auto first = net::buffer_sequence_begin(sr);
+                auto last = net::buffer_sequence_end(sr);
+                for (; first != last; ++first)
+                {
+                    auto buf = *first;
+                    auto f = reinterpret_cast<const char *>(buf.data());
+                    auto l = f + buf.size();
+                    result.insert(std::end(result), f, l);
```
Comment (vinniefalco) [inline test/beast/core/buffers_adaptor.cpp]: Why `std::end`? The type of `result` is known, this is not a generic context. Nothing is gained by calling `std::end` instead of just using `result.end`. Or even calling `result.append(f, l)`. I would like to understand why you chose one over the other, lets discuss.

Comment (madmongo1) [reply test/beast/core/buffers_adaptor.cpp]: I use std::end by force of habit, since it works in all contexts so I don't have to waste time reasoning about the underlying container. It's a zero cost abstraction, and I prefer free functions over member functions.
I could have used append, but it's specific to std::string and not generic across all containers, so to me it smells bad because it makes the code logic brittle (if I choose a different container type, I have to re-write logic, which makes no sense).
